packages/eslint-plugin/src/rules/no-misused-spread.ts

```diff
@@ -0,0 +1,201 @@
+            description:
+              'Whether to allow spreading class instances in objects.',
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+
+  defaultOptions: [
+    {
+      allowClassInstances: false,
+    },
+  ],
+
+  create(context, [options]) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function checkArraySpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isString(type)) {
+        context.report({
+          node,
+          messageId: 'noStringSpreadInArray',
+        });
+
+        return;
+      }
+    }
+
+    function checkObjectSpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isArray(type, checker)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Array',
+          },
+        });
+
+        return;
+      }
+
+      if (isPromise(services.program, type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Promise',
+          },
+        });
+
+        return;
+      }
+
+      if (isFunctionWithoutProps(type)) {
+        context.report({
+          node,
+          messageId: 'noFunctionSpreadInObject',
+        });
+
+        return;
+      }
+
+      if (isIterable(type, checker) && !isString(type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Iterable',
+          },
+        });
+
+        return;
+      }
+
+      if (!options.allowClassInstances && isClassInstance(type)) {
+        context.report({
+          node,
+          messageId: 'noClassInstanceSpreadInObject',
+        });
+
+        return;
+      }
+
+      if (node.argument.type === AST_NODE_TYPES.ClassExpression) {
+        context.report({
+          node,
+          messageId: 'noClassDeclarationSpreadInObject',
+        });
+
+        return;
+      }
+    }
+
+    return {
+      'ArrayExpression > SpreadElement': checkArraySpread,
+      'ObjectExpression > SpreadElement': checkObjectSpread,
+    };
+  },
+});
+
+function isArray(type: ts.Type, checker: ts.TypeChecker): boolean {
+  if (type.isUnion() || type.isIntersection()) {
+    return type.types.some(t => isArray(t, checker));
+  }
+
+  return checker.isArrayType(type) || checker.isTupleType(type);
+}
+
+function isIterable(type: ts.Type, checker: ts.TypeChecker): boolean {
+  return tsutils
+    .typeParts(type)
+    .some(
+      t => !!tsutils.getWellKnownSymbolPropertyOfType(t, 'iterator', checker),
+    );
+}
+
+function isString(type: ts.Type): boolean {
+  return tsutils
+    .typeParts(type)
+    .some(t => !!isTypeFlagSet(t, ts.TypeFlags.StringLike));
+}
```

Comment:
```suggestion
function isString(type: ts.Type): boolean {
  return tsutils
    .unionTypeParts(type)
    .some(unionPart => tsutils.intersectionTypeParts(unionPart)
      .some(intersectionPart => !!isTypeFlagSet(t, ts.TypeFlags.StringLike))
    );
}
```

`typeParts` iterates through the first level of `.types` property only. Looks like a bug:

```ts
declare const aa: (string & { foo: 1 }) | (string & { bar: 1 })
[...aa] // not reported - bug

declare const bb: (string & { foo: 1 })
[...bb] // reported - ok
```

[playground](https://deploy-preview-8509--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3lKAueAFJjAJYoDm8AZPAN7wBmSS%2BAjPAL4CU8APocTKUa9AEaw2nLgCgA2gDpFOALrwA9GvgokWOAAckMDCGDwAtPFEBXctOmhIsBMnRZRo-EQykK1Oo2ZJbjlFeXdVDXh9Q2NTCyQAayA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1kxaIgCGQ9FETRoHaJHBgAviDVA&tsconfig=&tokens=false)

Reply:
Great catch!